design-proposals/default-instancetypes-from-volumes.md

```diff
@@ -0,0 +1,86 @@
+# Overview
+
+[Instancetypes and preferences](https://kubevirt.io/user-guide/virtual_machines/instancetypes/) allow a users to pick a set of predefined resource, performance and other runtime preferences for their `VirtualMachine` without having to worry about the complexity of KubeVirt's `DomainSpec`.
+
+This design document aims to outline a simple way this feature could be further extended by allowing users to opt into having a default instancetype and/or preference inferred from a given `Volume` associated with their `VirtualMachine`.
+
+## Motivation
+
+The inference of a default instancetype and preference from a `Volume` associated with the `VirtualMachine` further reduces the choices a user has to make in order to get a running `VirtualMachineInstance` to a singular choice about which runnable image they would like to boot.
+
+## Goals
+
+* Allow users to opt into having a default instancetype and preference inferred from a given `Volume` associated with their `VirtualMachine`.
+
+## Non Goals
+
+* Automatic annotation of a `Volume` or underlying type (such as a `PVC`) is not covered as part of this work but is something that will be investigated in the future.
+
+## User Stories
+
+* As an admin, user or third party creator of a `Volume` (for example the `CDI` project using `DataVolumes` and `PVCs`),  I want to provide annotations that suggest the default instancetype or set of preferences.
+
+* As a user, I want to opt into allowing KubeVirt to infer these defaults from a volume associated with my VirtualMachine.
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+The `Volume` type will be extended to include a new `inferInstancetype` and `inferPreference` boolean attributes that will control this behaviour, defaulting to `False`.
+
+The attribute will only be allowed against a single `Volume` associated with a `VirtualMachine`, with this behaviour enforced early within the API by the mutation webhook before any additional logic is executed.
+
+When enabled against a single `Volume` KubeVirt will then look at the underlying `Volume` type (such as a `PVC`) for the following annotations:
+
+* `instancetype.kubevirt.io/defaultInstancetype`
+* `instancetype.kubevirt.io/defaultInstancetypeKind` (Defaults to `VirtualMachineClusterInstancetype`)
+* `instancetype.kubevirt.io/defaultPreference`
+* `instancetype.kubevirt.io/defaultPreferenceKind` (Defaults to `VirtualMachineClusterPreference`)
+
+If found these annotations will be used to populate the `InstancetypeMatcher` and `PreferenceMatcher` of the `VirtualMachine` by the mutation webhook with the existing checks of the validation webhook then used to ensure these resources exist and apply without conflicts to the `VirtualMachine` and eventual `VirtualMachineInstance` at runtime.
```

Comment:
@lyarwood When is the mutation webhook called? Are already existing attributes of the VM overwritten?

Reply:
> @lyarwood When is the mutation webhook called? Are already existing attributes of the VM overwritten?

First and no, nothing already defined would be overwritten.

Reply:
If we move these to the `{Instancetype,Preference}Matcher` then I'd rather they stay to explain where the initial values were taken from.

Reply:
I do not agree. I would prefer if the yaml/object model would describe the current state without inconsistencies over that it explains his  own history.

Reply:
The usual flow of creating a VM is
1. start with a spec.dataVolumeTemplates
2. the dataVolumeTemplates will be cloned
3. the VM will be booted the first time

Can the preferences and instancetypes inferred from the dataVolumeTemplates? This would help to have a clear point in time when the inferring would happen and avoid the inconsistency this way.

Reply:
> I do not agree. I would prefer if the yaml/object model would describe the current state without inconsistencies over that it explains his own history.

Okay that's understandable, we can always drop this in the mutation webhook while applying the changes to the matchers.



> The usual flow of creating a VM is
> 
> 1. start with a spec.dataVolumeTemplates
> 2. the dataVolumeTemplates will be cloned
> 3. the VM will be booted the first time
> 
> Can the preferences and instancetypes inferred from the dataVolumeTemplates? This would help to have a clear point in time when the inferring would happen and avoid the inconsistency this way. In this case, this document should mention that during the cloning of the dataVolumeTemplates the annotations should be not cloned.

This really comes down to what the source of the DataVolumeTemplate is and if it support annotations. If it's a PVC then yes we should be able to easily look these up but that wouldn't be possible for http etc.

Reply:
> This really comes down to what the source of the DataVolumeTemplate is and if it support annotations. If it's a PVC then yes we should be able to easily look these up but that wouldn't be possible for http etc.

That's fine. While http provides at least the size of the disk ( via qemu-img info ), the question how to provide information about a containerdisk is still open.

Reply:
> the question how to provide information about a containerdisk is still open.

contains can have labels associated with them. a tool like skopeo can be used to retrieve those labels without actually pulling the image.

For example, with openshift i can get the label representing the release version from the a container image. This is kind of a nonsensical example since the release version is in the image's tag as well, but it shows what I'm talking about.

```
skopeo inspect --config docker://quay.io/openshift-release-dev/ocp-release:4.10.12-x86_64 | grep \"io.openshift.release\"
            "io.openshift.release": "4.10.12",
```

Skopeo is written in golang. So it's fairly trivial to use that logic internally within our controllers if we need to.

Reply:
Yeah indeed container disks seem like another easy win here, I haven't listed them in the doc yet but given the above it should be trivial enough.